satpy/readers/seviri_base.py

```diff
@@ -372,14 +378,90 @@ def _tl15(self, data, wavenumber):
         """
         reflectance = np.pi * data * 100.0 / solar_irradiance
-        return apply_earthsun_distance_correction(reflectance, self.start_time)
+        return apply_earthsun_distance_correction(reflectance, self._scan_time)
+
+
+class SEVIRICalibrationHandler:
+    """Calibration handler for SEVIRI HRIT-, native- and netCDF-formats.
+
+    Handles selection of calibration coefficients and calls the appropriate
+    calibration algorithm.
+    """
+
+    def __init__(self, platform_id, channel_name, coefs, calib_mode, scan_time):
+        """Initialize the calibration handler."""
+        self._platform_id = platform_id
+        self._channel_name = channel_name
+        self._coefs = coefs
+        self._calib_mode = calib_mode.upper()
+        self._scan_time = scan_time
+        self._algo = SEVIRICalibrationAlgorithm(
+            platform_id=self._platform_id,
+            scan_time=self._scan_time
+        )
+
+        valid_modes = ('NOMINAL', 'GSICS')
+        if self._calib_mode not in valid_modes:
+            raise ValueError(
+                'Invalid calibration mode: {}. Choose one of {}'.format(
+                    self._calib_mode, valid_modes)
+            )
+
+    def calibrate(self, data, calibration):
+        """Calibrate the given data."""
+        if calibration == 'counts':
+            res = data
+        elif calibration in ['radiance', 'reflectance',
+                             'brightness_temperature']:
+            gain, offset = self.get_gain_offset()
+            res = self._algo.convert_to_radiance(
+                data.astype(np.float32), gain, offset
+            )
+        else:
+            raise ValueError(
+                'Invalid calibration {} for channel {}'.format(
+                    calibration, self._channel_name
+                )
+            )
+
+        if calibration == 'reflectance':
+            solar_irradiance = CALIB[self._platform_id][self._channel_name]["F"]
+            res = self._algo.vis_calibrate(res, solar_irradiance)
+        elif calibration == 'brightness_temperature':
+            res = self._algo.ir_calibrate(
+                res, self._channel_name, self._coefs['radiance_type']
+            )
+
+        return res
+
+    def get_gain_offset(self):
+        """Get gain & offset for calibration from counts to radiance.
+
+        Choices for internal coefficients are nominal or GSICS. External
+        coefficients take precedence.
+        """
+        coefs = self._coefs['coefs']
+
+        # Select internal coefficients for the given calibration mode
+        if self._calib_mode != 'GSICS' or self._channel_name in VIS_CHANNELS:
+            # GSICS doesn't have calibration coeffs for VIS channels
+            internal_gain = coefs['NOMINAL']['gain']
+            internal_offset = coefs['NOMINAL']['offset']
+        else:
+            internal_gain = coefs['GSICS']['gain']
+            internal_offset = coefs['GSICS']['offset'] * internal_gain
```

Comment:
In theory the VIS channels also have GSICS coefficients, they just happen to be empty at the moment. I think there's a plan to implement VIS GSICS sometime in 2021, though.
As such, maybe it'd be better to remove the `self._channel_name in VIS_CHANNELS` and instead check if `coefficient == 0`?

Reply:
Oh, I didn't know that! That's a good idea then!

Reply:
That made the code more robust, thanks for the suggestion!